### ISSUE TITLE
[Accepted] SDL 0044 - Use Boost library in SDL

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#133]
Hello SDL community,

The review of the revised proposal "Use Boost library in SDL" begins now and runs through May 2, 2017. The original review of "Use Boost library in SDL" occurred March 28, 2017 through April 4, 2017. The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0044-use-Boost-library.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/133

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by joeljfischer** [comment]
1. This is a difficult proposal to understand grammatically and has a number of spelling errors. I believe the committee should consider how clear proposals need to be in order to be accepted.

2. I'm not certain, but this appears to be dependent on SDL-0043. In my research, the boost library is only available at C++11. This proposal should not have been simultaneously in review with it.

3. Although this proposal doesn't state it, the Boost library's license appears to be MIT. Before this would be accepted, I think this would have to be cleared by the SDLC lawyers. The MIT is very permissive, so it should be okay.

With that said, the overall idea of this proposal seems good. Having well-tested library support for common functions that are broadly cross-platform seems like a good addition. If a platform is not supported however, will having this library be more of a burden than our own custom code?

**Comment by AGaliuzov** [comment]
@joeljfischer 

> I'm not certain, but this appears to be dependent on SDL-0043. In my research, the boost library is only available at C++11. This proposal should not have been simultaneously in review with it.

The boost library available for c++98 as well. Basically, it has different parts(sub-libraries), some of them indeed are available only together with c++11 whereas other are working fine with c++98.

Additional note: since version 1.56 Boost library officially supports QCC4.2 which is default compiler for QNX6.5 and based on gcc4.2. The gcc4.2 does not support c++11.

Thus, I don't see any dependency on SDL-0043 at all.

**Comment by joeygrover** [comment]
#### Overall
- I support the use of boost libraries in the core project. Boost is a well established library with high quality code and a permissive license. 
- As stated, the amount of features we could leverage will depend on if we move to c++11 or stay at c++98, but there are enough features at the c++98 level that make this addition worthwhile. 

#### Concerns
- If this is accepted, we will have to weigh out if we wish to refactor the different parts of the project to use the Boost API's over our custom classes.
- If we don't believe refactoring certain code with the Boost API's would be worthwhile, we will need to consider if we should always use our custom implementation of those specific classes/features. We should avoid getting into situations where sometimes we use one and sometimes we use the other. 
- If the steering committee accepts this proposal we will need to then take it through the SDLC legal team to ensure we can add it to the project.

**Comment by joeljfischer** [comment]
Is there any sort of guess on timeline or lines of code changed for this?

**Comment by joeygrover** [comment]
I believe it would depend on which aspect you wanted to attempt to refactor to use the Boost APIs. Maybe like, refactor to use Boosts Async features over custom Threads as an example?

**Comment by JackLivio** [comment]
I agree with Joey and Joel's comments. Would also be interested in a timeline or an estimate for the amount of work for these changes. I am under the impression that the amount of refactoring to make these changes worthwhile will be astronomical.

**Comment by AGaliuzov** [comment]
@JackLivio  @joeljfischer @joeygrover 
The main idea in this proposal is to describe what is boost, what advantages it has and why it could be useful for SDL.
If committee will accept this proposal and allow to use boost in SDL we are going to prepare several proposals for the future refactoring. 

To me, it looks like following
1. Substitute own implementation of the `FileSystem`  and `DateTime` classes in SDL
2. Remove own timer's, which is not sufficient enough, in favor of the boost's
3.  Use Asio for the threads management and async operations - this one could be split to several small pieces 
4. Use Boost's synchronization primitives

**Comment by theresalech** [comment]
The Steering Committee has deferred this proposal, as they've agreed the author will need to prepare a more specific proposal, with which exact components could be replaced with the Boost library.  The revised proposal should also clarify which specific version of the Boost library would be used.

**Comment by AGaliuzov** [comment]
The pull request for the additional proposal is located at the following link
https://github.com/smartdevicelink/sdl_evolution/pull/148

**Comment by AGaliuzov** [comment]
The proposed version of the Boost library of **1.56**

**Comment by tpulatha** [comment]
Can I ask why 1.56 as this is a quite old version? (1.64 being the newest)

**Comment by AGaliuzov** [comment]
@tpulatha  This is the first version which is officially supported QCC4 for QNX6.5. and almost all its features are working with the old compilers. The newest versions of the boots library contain some number of sub-libraries which could be compiled only with newest c++ compiler as they required some specific c++11/14 features.

In order to avoid confusion about features set between hardware which uses legacy QNX and regular Linux(like Ubuntu 16) I proposed to start at least from 1.56 to be able to use the same version under all systems we have to support.

**Comment by joeljfischer** [comment]
I must admit to still being leery about this. It seems like a lot of work for relatively small gains. As I am not a Core developer, however, I will defer to others' judgment.

**Comment by AGaliuzov** [comment]
@joeljfischer As I highlighted in the proposal the main goal is to substitute our own implementation of the different algorithms and primitives we are currently using in SDL with some stable and production ready solution which will not require any efforts for re-working and/or extension during porting to any other platform. As the example, I show the `FileSystem` class where we way 
1. Reduce code amount
2. Make the code more stable and error free
3. Simplify the portability

So the advantage is quite obvious in this case - if we have a good framework which provides us ready functionality, we have to prefer to use it instead of writing own implementation for the same

**Comment by theresalech** [comment]
The Steering Committee has noted that there may be overlapping benefits of Boost and C++11 once C++11 has been implemented.  However, as a date for C++11 implementation has not yet been determined, the benefits of using Boost library in the interim warrant this proposal's acceptance.

**Comment by theresalech** [comment]
Issue entered: [[SDL 0044] Use Boost library in SDL](https://github.com/smartdevicelink/sdl_core/issues/1523)
